cartographer_ros/cartographer_ros/pbstream_to_ros_map_main.cc

```diff
@@ -0,0 +1,133 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <map>
+#include <string>
+
+#include "cartographer/io/proto_stream.h"
+#include "cartographer/io/submap_painter.h"
+#include "cartographer/mapping/proto/pose_graph.pb.h"
+#include "cartographer/mapping/proto/serialization.pb.h"
+#include "cartographer/mapping/proto/submap.pb.h"
+#include "cartographer/mapping/submaps.h"
+#include "cartographer/mapping_2d/probability_grid.h"
+#include "cartographer/mapping_2d/submaps.h"
+#include "cartographer/mapping_3d/submaps.h"
+#include "cartographer_ros/ros_map.h"
+#include "cartographer_ros/submap.h"
+#include "gflags/gflags.h"
+#include "glog/logging.h"
+
+DEFINE_string(pbstream_filename, "",
+              "Filename of a pbstream to draw a map from.");
+DEFINE_string(map_filestem, "map", "Stem of the output files.");
+DEFINE_double(resolution, 0.05, "Resolution of a grid cell in the drawn map.");
+
+namespace cartographer_ros {
+namespace {
+
+void FillSubmapState(const ::cartographer::transform::Rigid3d& submap_pose,
+                     const ::cartographer::mapping::proto::Submap& proto,
+                     ::cartographer::io::SubmapSlice* submap_slice) {
+  ::cartographer::mapping::proto::SubmapQuery::Response response;
+  ::cartographer::transform::Rigid3d local_pose;
+  if (proto.has_submap_3d()) {
+    ::cartographer::mapping_3d::Submap submap(proto.submap_3d());
+    local_pose = submap.local_pose();
+    submap.ToResponseProto(submap_pose, &response);
+  } else {
+    ::cartographer::mapping_2d::Submap submap(proto.submap_2d());
+    local_pose = submap.local_pose();
+    submap.ToResponseProto(submap_pose, &response);
+  }
+  submap_slice->pose = submap_pose;
+
+  auto& texture_proto = response.textures(0);
+  const SubmapTexture::Pixels pixels = UnpackTextureData(
+      texture_proto.cells(), texture_proto.width(), texture_proto.height());
+  submap_slice->width = texture_proto.width();
+  submap_slice->height = texture_proto.height();
+  submap_slice->resolution = texture_proto.resolution();
+  submap_slice->slice_pose =
+      ::cartographer::transform::ToRigid3(texture_proto.slice_pose());
+  submap_slice->surface =
+      DrawTexture(pixels.intensity, pixels.alpha, texture_proto.width(),
+                  texture_proto.height(), &submap_slice->cairo_data);
+}
+
+void Run(const std::string& pbstream_filename, const std::string& map_filestem,
+         const double resolution) {
+  ::cartographer::io::ProtoStreamReader reader(pbstream_filename);
+
+  // Skip Pose graph
```

Comment:
Full sentences end with a period. Similar below. Also typo: `pose`.

Reply:
removed that comment. It did not add to the narrative of the code.